flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java

```diff
@@ -526,7 +545,8 @@ private static CPUResource getCpuCores(final Configuration config) {
 			}
 			return new CPUResource(cpuCores);
 		} else {
-			return new CPUResource(config.getInteger(TaskManagerOptions.NUM_TASK_SLOTS));
+			final BigDecimal defaultSlotFraction =  getDefaultSlotFraction(config);
+			return new CPUResource(BigDecimal.valueOf(1.0).divide(defaultSlotFraction, 16, RoundingMode.DOWN).doubleValue());
```

Comment:
True, we should avoid the back and forth conversions.

And for `CPUResource`, I guess what you mean is to avoid having public constructors and factory methods at the same time? We need a private constructor anyway if we want to use factory methods.

And I'm not sure whether we need to describe the value type in the name of factory methods. We can simply have factory methods with different argument type, like `create(BigDecimal)` and `create(double)`. So the factory of double type stays the same name as the overridden methods in its super class.

Reply:
True, we should avoid the back and forth conversions.

I would not say `CPUResource#create` is a factory method. The method is not static, and the reason we cannot make it static is that we use it in `Resource#merge/subtract/multiply` to create new instances of `Resource`'s subclasses without caring about the `name`.

`CPUResource#create` is only used internally by its superclass `Resource`. Externally, only constructors are used. And I think here we can simply make `CPUResource(BigDecimal)` public to avoid the back and forth conversions.

WDYT?

Reply:
True, the factory is the interface implementation. I agree we can promote the constructor.

Reply:
Another thing, do we actually need `cpuCores` to be of `CPUResource` type in `TaskExecutorResourceSpec`? If we do not use the fact that it is `CPUResource`, we should declare with `Resource` type.

Reply:
As discussed offline, we decided to keep the type `CPUResource`, to avoid accidentally assigning other type of `Resource` (e.g., `GPUResource`) to this variable.